### PR TITLE
Adds interpolation to the science sensor refactor

### PR DESCRIPTION
**Pull request by arjo129**
Requires [igntionrobotics/ign-math#412](https://github.com/ignitionrobotics/ign-math/pull/412)

Signed-off-by: Arjo Chakravarty <arjo@openrobotics.org>

### CODE REVIEW

File: lrauv_ignition_plugins/src/ScienceSensorsSystem.cc [hunk 1]
```diff
@@ -685,57 +685,83 @@ void ScienceSensorsSystem::PostUpdate(const ignition::gazebo::UpdateInfo &_info,
...
+      /// igndbg << "Got" << interpolators.size() << " interpolators" << std::endl;
+      /// for (auto interp: interpolators)
+      /// {
+      ///   if (!interp.index.has_value()) continue;
+      ///   igndbg << "Chlorophyll sensor: " <<
+      ///     this->dataPtr->chlorophyllArr[this->dataPtr->timeIdx][
+      ///       interp.index.value()] << "@" << interp.position << std::endl;
+      ///   igndbg << "My distance is " << interp.position.X() - sphericalDepthCorrected.X()
+      ///   << ", " << interp.position.Y() - sphericalDepthCorrected.Y() << std::endl;
+      /// }
```
Comment (hidmic) [inline lrauv_ignition_plugins/src/ScienceSensorsSystem.cc:738]: @arjo129 debugging leftover?

File: lrauv_ignition_plugins/test/test_sensor.cc [hunk 2]
```diff
@@ -51,43 +107,40 @@ TEST(SensorTest, Sensor)
   auto fixture = std::make_unique<ignition::gazebo::TestFixture>(
       ignition::common::joinPaths(
       std::string(PROJECT_SOURCE_PATH), "worlds", "empty_environment.sdf"));
-
-  ignition::gazebo::Entity vehicle1{ignition::gazebo::kNullEntity};
   unsigned int iterations{0u};
-  std::vector<ignition::math::Pose3d> poses1;
-  std::vector<ignition::math::Vector3d> latLon1;
+
+  std::atomic<bool> spawnedAllVehicles = {false};
```
Comment (hidmic) [inline lrauv_ignition_plugins/test/test_sensor.cc:112]: @arjo129 fixture runs synchronously, why an atomic?

File: lrauv_ignition_plugins/test/test_sensor.cc [hunk 3]
```diff
@@ -99,58 +152,44 @@ TEST(SensorTest, Sensor)
...
+
+  // In the middle of a quadrant
+  SpawnVehicle(
+    spawnPub, "vehicle4", 36.7999992370605, -122.7, 48, 0);
+  node.Subscribe("/model/vehicle4/chlorophyll", &ChlorophyllVeh4Cb);
 
   // Check that vehicle was spawned
   int expectedIterations = iterations;
-  for (sleep = 0; ignition::gazebo::kNullEntity == vehicle1 && sleep < maxSleep;
+  for (sleep = 0; spawnedAllVehicles.load() == false && sleep < maxSleep;
```
Comment (hidmic) [inline lrauv_ignition_plugins/test/test_sensor.cc:178]: @arjo129 nit^2: 
```suggestion
  for (sleep = 0; !spawnedAllVehicles.load() && sleep < maxSleep;
```
is equivalent.

File: lrauv_ignition_plugins/test/test_sensor.cc [hunk 4]
```diff
@@ -99,58 +152,44 @@ TEST(SensorTest, Sensor)
...
-
-  // Higher tolerance for lat/lon because of the conversions
-  double latLonTol{2e-2};
-
-  fixture->Server()->Run(true, 10000, false);
+  fixture->Server()->Run(true, 10, false);
 
+  for (auto msg: received_msg)
+  {
+    EXPECT_TRUE(msg);
```
Comment (hidmic) [inline lrauv_ignition_plugins/test/test_sensor.cc:193]: @arjo129 meta: this is racing with Ignition Transport. Consider using a timeout (measured against steady or system clocks) instead. 

Writing this kind of tests will get easier once #195 lands.

Comment (arjo129) [reply lrauv_ignition_plugins/test/test_sensor.cc:193]: Good point. I think we could wait till #195 goes in before I rewrite this.

**Comment by arjo129** [comment]
I'll rebase this. Seem to have messed it up

File: lrauv_ignition_plugins/src/ScienceSensorsSystem.cc [hunk 5]
```diff
@@ -1498,232 +666,118 @@ void ScienceSensorsSystem::PostUpdate(const ignition::gazebo::UpdateInfo &_info,
...
-            << std::round(nbrPt.y * 1000) / 1000.0 << ", "
-            << std::round(nbrPt.z * 1000) / 1000.0
-            << "), squared distance " << spatialSqrDist[i]
-            << " m" << std::endl;
-        }
-      }
-      reinterpolate = true;
+    IGN_PROFILE("ScienceSensorsSystem::Interpolation");
+    if (interpolators.size() == 0) return;
+    if (!interpolators[0].index.has_value()) return;
```
Comment (hidmic) [inline lrauv_ignition_plugins/src/ScienceSensorsSystem.cc:690]: @arjo129 shall we set NaNs even when interpolation is not possible too?

File: lrauv_ignition_plugins/src/ScienceSensorsSystem.cc [hunk 6]
```diff
@@ -1498,232 +666,118 @@ void ScienceSensorsSystem::PostUpdate(const ignition::gazebo::UpdateInfo &_info,
...
-          igndbg << "Interpolating temperature" << std::endl;
-        this->dataPtr->ExtractElements(
-          this->dataPtr->temperatureArr[this->dataPtr->timeIdx], spatialIdx,
-          interpolationValues);
-        float temp = this->dataPtr->InterpolateData(
-          sensorPosENUEigen, interpolatorXYZsMat, interpolationValues);
-
         ignition::math::Temperature tempC;
-        tempC.SetCelsius(temp);
+        tempC.SetCelsius(temp.value());
```
Comment (hidmic) [inline lrauv_ignition_plugins/src/ScienceSensorsSystem.cc:717]: @arjo129 nit^N!: `value_or()` may be practical here to deduplicate some code.

**Review by hidmic — APPROVED** [review]
Not much to add. Previous comments are minor, so I'm OK with moving forward as-is if necessary for the demo.

**Pull request merged**